File: lib/monoboly_deal_web/channels/game_channel.ex
defmodule MonobolyDealWeb.GameChannel do
  use MonobolyDealWeb, :channel

  alias MonobolyDeal.Game.Server
  alias MonobolyDealWeb.Presence

  def join("games:" <> game_name, _params, socket) do
    case Server.game_pid(game_name) do
      pid when is_pid(pid) ->
        case Server.join(game_name, current_player(socket)) do
          {:ok, _game_state} ->
            send(self(), {:after_join, game_name})
            {:ok, assign(socket, :game_name, game_name)}

          {:error, error} ->
            {:error, error}
        end

      nil ->
        {:error, %{error: :game_not_found, message: "Oops! We couldn't find that game."}}
    end
  end

  def handle_info({:after_join, game_name}, socket) do
    game_state = Server.game_state(game_name)
    broadcast(socket, "game_state", game_state)

    player_state = Server.player_state(game_name, current_player(socket))
    push(socket, "player_state", player_state)

    push(socket, "presence_state", Presence.list(socket))

    {:ok, _} =
      Presence.track(
        socket,
        current_player(socket).name,
        %{
          online_at: inspect(System.system_time(:seconds))
        }
      )

    {:noreply, socket}
  end

  def handle_in("deal_hand", _params, socket) do
    "games:" <> game_name = socket.topic

    case Server.game_pid(game_name) do
      pid when is_pid(pid) ->
        Server.deal_hand(game_name)

        game_state = Server.game_state(game_name)
        push(socket, "game_state", game_state)
        {:noreply, socket}

      nil ->
        {:reply, {:error, %{reason: "Game does not exist"}}, socket}
    end
  end

  defp current_player(socket) do
    socket.assigns.current_player
  end
end

File: assets/js/communication/GameChannel.ts
import {Channel, Socket} from "phoenix"
import {GameState} from "../models/GameState"
import {Player} from "../models/Player"
import {PlayerSerializer} from "../serializers/PlayerSerializer"

export class GameChannel {
  public channel: Channel
  private playerSerializer: PlayerSerializer

  constructor(gameName: string, socket: Socket) {
    this.channel = socket.channel("games:" + gameName)
    this.playerSerializer = new PlayerSerializer()

    this.channel
      .join()
      .receive("ok", (response: any) => {
        console.log("game channel joined", {response})
      })
      .receive("error", (error: any) => {
        console.log("game channel join failed", {error})
      })
  }

  public onGameStateUpdated(callBack: (gameState: GameState) => void) {
    this.channel.on("game_state", (response: any) => {
      if (response) {
        const gameState = new GameState({
          gameStarted: response.started,
          name: response.game_name,
          players: response.players.map((player: any) => {
            return this.playerSerializer.deserialize(player)
          }),
        })

        callBack(gameState)
      }
    })
  }

  public onPlayerStateUpdated(callBack: (playerState: Player) => void) {
    this.channel.on("player_state", (response: any) => {
      if (response) {
        const player = this.playerSerializer.deserialize(response)
        callBack(player)
      }
    })
  }

  public dealHand() {
    this.channel
      .push("deal_hand", {})
      .receive("error", (error) => {
        console.log("deal_hand failed", {error})
      })
  }
}
